### PR TITLE
Support for repeatable directives, refactored schema visitors (stateless), directives validation, small optimizations

### LINKED ISSUE #1496: Support for repeatable directives
graphql/graphql-spec#472 requires changes in:
1. Parser
2. Validation
3. Introspection
4. Some internal logic

### PR DESCRIPTION
**Pull request by sungam3r**
This is a continuation of working with directives. Fixes #1496.

### CODE REVIEW

**Comment by sungam3r** [comment]
@Shane32 This PR is relatively simple. It develops ideas for which the preparatory work has already been done earlier - here, in parser repo, spec repo.

**Comment by sungam3r** [comment]
I did not make edits in the migration guide so as not to generate conflicts.

File: src/GraphQL.Tests/Introspection/AppliedDirectivesTests.cs [hunk 1]
```diff
@@ -52,7 +52,7 @@ private class TraitsDirective : DirectiveGraphType
             public override bool? Introspectable => true;
 
             public TraitsDirective()
-                : base("traits", DirectiveLocation.Schema | DirectiveLocation.Object | DirectiveLocation.FieldDefinition)
```
Comment (sungam3r) [inline src/GraphQL.Tests/Introspection/AppliedDirectivesTests.cs:55]: 🤦

File: src/GraphQL.Tests/Types/InvalidEnumGraphTypeTests.cs [hunk 2]
```diff
@@ -13,7 +13,18 @@ private enum Invalid
         }
 
         [Fact]
-        public void AddValue_whenEnumContainsInvalidCharacters_shouldThrowArgumentException() =>
-            Should.Throw<ArgumentOutOfRangeException>(() => new EnumerationGraphType<Invalid>());
+        public void AddValue_whenEnumContainsInvalidCharacters_shouldThrowArgumentException()
+        {
+            // race condition with does_not_throw_with_filtering_nameconverter test
```
Comment (sungam3r) [inline src/GraphQL.Tests/Types/InvalidEnumGraphTypeTests.cs:18]: one more test for invalid names

File: src/GraphQL/Types/Collections/SchemaTypes.cs [hunk 3]
```diff
@@ -160,7 +160,7 @@ private void CheckSealed()
             Func<Type, IGraphType> resolveType,
             ISchema schema)
         {
-            var lookup = new SchemaTypes(schema.NameConverter ?? CamelCaseNameConverter.Instance, schema.Features.AppliedDirectives);
+            var lookup = new SchemaTypes(schema.NameConverter ?? CamelCaseNameConverter.Instance, schema.Features.AppliedDirectives, schema.Features.RepeatableDirectives);
```
Comment (sungam3r) [inline src/GraphQL/Types/Collections/SchemaTypes.cs:163]: Maybe it's time to pass a single `schema` argument?

File: src/GraphQL/Utilities/SchemaBuilder.cs [hunk 4]
```diff
@@ -460,24 +460,25 @@ protected virtual EnumerationGraphType ToEnumerationType(GraphQLEnumTypeDefiniti
...
             {
                 Description = directiveDef.Comment?.Text.ToString(),
+                Repeatable = directiveDef.Repeatable,
                 Arguments = ToQueryArguments(directiveDef.Arguments)
             };
-        }
 
-        private DirectiveLocation ToDirectiveLocation(string name)
-        {
-            var enums = new __DirectiveLocation(); //TODO: remove new, also see TODO above
```
Comment (sungam3r) [inline src/GraphQL/Utilities/SchemaBuilder.cs:473]: No more alocations here and above. Two TODOs are done.

**Comment by Shane32** [comment]
> I did not make edits in the migration guide so as not to generate conflicts.

I appreciate the thought, but it would actually be better if you added the notes so I have something to work from.  I’ll handle the merge. Or add some notes in the PR of the migration guide so I don’t forget anything.

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 3 above)
Comment (Shane32) [reply src/GraphQL/Types/Collections/SchemaTypes.cs:163]: Sure. I didn’t realize you changed it back previously. Please make it require ISchema, not Schema.

**Comment by Shane32** [comment]
I’m not at a pc now but I noticed a lot of references to Schema instead of ISchema. Unless the class is only used by the schema builder, I would expect that it should be ISchema. Can we change that?

**Comment by sungam3r** [comment]
Done. I myself will write on a separate documentation page everything that is done with the directives. I will also make one or more PRs with documentation and examples. Directives are the area that is required in my work, so I considered it right to finish what I have long wanted.

File: src/GraphQL/Types/Collections/SchemaTypes.cs (hunk 3 above)
Comment (sungam3r) [reply src/GraphQL/Types/Collections/SchemaTypes.cs]: Done.

**Comment by Shane32** [comment]
Thank you!!!

**Review by Shane32 — APPROVED** [review]

**Pull request merged**